framework/src/physics/DiffusionPhysicsBase.C

```diff
@@ -124,15 +127,53 @@ void
 DiffusionPhysicsBase::addInitialConditions()
 {
   InputParameters params = getFactory().getValidParams("FunctionIC");
-  assignBlocks(params, _blocks);
 
-  // always obey the user specification of initial conditions
+  // Get the list of blocks that have ics from components
+  std::vector<SubdomainName> component_ic_blocks;
+  for (const auto & comp_pair : _components_initial_conditions)
+  {
+    if (!comp_pair.second.count(_var_name))
+      continue;
+    const auto & comp_blocks = getActionComponent(comp_pair.first).blocks();
+    component_ic_blocks.insert(component_ic_blocks.end(), comp_blocks.begin(), comp_blocks.end());
+  }
+
+  // Keep only blocks that have no component IC
+  std::vector<SubdomainName> remaining_blocks;
+  for (const auto & block : _blocks)
+    if (std::find(component_ic_blocks.begin(), component_ic_blocks.end(), block) ==
+        component_ic_blocks.end())
+      remaining_blocks.push_back(block);
+  assignBlocks(params, remaining_blocks);
+
+  // first obey any component-specific initial condition
+  // then obey the user specification of initial conditions
+  // NOTE: we may conflict with ICs in the input
   // there are no default initial conditions
-  if (parameters().isParamSetByUser("initial_condition"))
+  mooseAssert(!parameters().hasDefault("initial_condition"), "Should not have a default");
+  if (parameters().isParamSetByUser("initial_condition") && remaining_blocks.size())
   {
     params.set<VariableName>("variable") = _var_name;
     params.set<FunctionName>("function") = getParam<FunctionName>("initial_condition");
 
     getProblem().addInitialCondition("FunctionIC", prefix() + _var_name + "_ic", params);
   }
 }
+
+void
+DiffusionPhysicsBase::addInitialConditionsFromComponents()
+{
+  InputParameters params = getFactory().getValidParams("FunctorIC");
+
+  for (const auto & comp_pair : _components_initial_conditions)
```

Comment:
You use both members of the pair, so let's use a structured binding for better readability

Reply:
done